### PR TITLE
scheduler: refactor operators.

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

File: server/schedule/operator.go [hunk 1]
```diff
@@ -14,88 +14,177 @@
...
+// CreateMovePeerOperator creates an Operator that replaces an old peer with a
+// new peer. It prevents removing leader by transfer its leadership first.
+func CreateMovePeerOperator(desc string, region *core.RegionInfo, kind core.ResourceKind, oldStore, newStore uint64, peerID uint64) *Operator {
+	if region.Leader != nil && region.Leader.GetStoreId() == oldStore {
+		newLeader := newStore
+		if follower := region.GetFollower(); follower != nil {
+			newLeader = follower.GetStoreId()
+		}
+		steps := []OperatorStep{
+			AddPeer{ToStore: newStore},
```
Comment (nolouch) [inline server/schedule/operator.go]: miss peer id

File: server/schedule/replica_checker.go [hunk 2]
```diff
@@ -54,15 +54,15 @@ func (r *ReplicaChecker) Check(region *core.RegionInfo) Operator {
 		if newPeer == nil {
 			return nil
 		}
-		return CreateAddPeerOperator(region, newPeer)
+		return NewOperator("makeUpReplica", region.GetId(), core.RegionKind, AddPeer{ToStore: newPeer.GetStoreId(), PeerID: newPeer.GetId()})
```
Comment (nolouch) [inline server/schedule/replica_checker.go]: should also extract a function?

File: server/coordinator.go [hunk 3]
```diff
@@ -285,94 +291,131 @@ func (c *coordinator) addOperator(op schedule.Operator) bool {
...
+		op := elem.Value.(*schedule.Operator)
+		if op.ResourceKind() == kind {
 			operators = append(operators, op)
 		}
 	}
 
 	return operators
 }
 
+func (c *coordinator) sendScheduleCommand(region *core.RegionInfo, step schedule.OperatorStep) {
```
Comment (nolouch) [inline server/coordinator.go]: add a log to descript the step

File: server/schedule/operator.go [hunk 4]
```diff
@@ -14,88 +14,181 @@
...
+func (o *Operator) RegionID() uint64 {
+	return o.regionID
+}
+
+// ResourceKind returns operator's resource kind.
+func (o *Operator) ResourceKind() core.ResourceKind {
+	return o.kind
+}
+
+// ElapseTime returns duration since it was created.
```
Comment (huachaohuang) [inline server/schedule/operator.go]: s/ElapseTime/ElapsedTime

File: server/schedule/operator.go [hunk 5]
```diff
@@ -14,88 +14,181 @@
...
+		return o.steps[i]
+	}
+	return nil
+}
+
+// Check checks if current step is finished, returns next step to take action.
+func (o *Operator) Check(region *core.RegionInfo) OperatorStep {
+	for step := atomic.LoadInt32(&o.currentStep); int(step) < len(o.steps); step++ {
+		if o.steps[int(step)].IsFinish(region) {
+			atomic.StoreInt32(&o.currentStep, step+1)
```
Comment (huachaohuang) [inline server/schedule/operator.go]: `atomic.AddInt32(&o.currentStep, 1)`?

Comment (disksing) [reply server/schedule/operator.go]: If 2 goroutines call Check() at the same time, `currentStep` could be incorrectly set to a larger value.

File: server/schedule/operator.go [hunk 6]
```diff
@@ -14,88 +14,181 @@
...
-	s, err := strconv.Unquote(string(text))
-	if err != nil {
-		return errors.Trace(err)
+// IsFinish checks if current step is finished.
+func (rp RemovePeer) IsFinish(region *core.RegionInfo) bool {
+	return region.GetStorePeer(rp.FromStore) == nil
+}
+
+// Operator contains execution steps generated by scheduler.
+type Operator struct {
```
Comment (huachaohuang) [inline server/schedule/operator.go]: Should add some tests for operators in this file.

File: server/schedule/operator.go (hunk 5 above)
Comment (huachaohuang) [reply server/schedule/operator.go]: Maybe add some comments, seems not straight forward.

**Pull request merged**